homeassistant/components/intellifire/__init__.py

```diff
@@ -32,79 +43,127 @@
+        new[CONF_AUTH_COOKIE] = new_data.auth_cookie
+
+        new[CONF_IP_ADDRESS] = new_data.ip_address
+        new[CONF_SERIAL] = new_data.serial
+
+        config_entry.version = 1
         hass.config_entries.async_update_entry(
-            entry,
-            data={
-                **entry.data,
-                CONF_API_KEY: api_key,
-                CONF_USER_ID: user_id,
-            },
+            config_entry,
+            data=new,
+            options={CONF_READ_MODE: "local", CONF_CONTROL_MODE: "local"},
+            unique_id=serial,
         )
+        LOGGER.debug("Pseudo Migration %s successful", config_entry.version)
 
-    else:
-        api_key = entry.data[CONF_API_KEY]
-        user_id = entry.data[CONF_USER_ID]
+    return config_entry
 
-    # Instantiate local control
-    api = IntellifireAPILocal(
-        fireplace_ip=entry.data[CONF_HOST],
-        api_key=api_key,
-        user_id=user_id,
-    )
 
-    # Define the update coordinator
-    coordinator = IntellifireDataUpdateCoordinator(
-        hass=hass,
-        api=api,
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up IntelliFire from a config entry."""
+    LOGGER.debug("Setting up config entry: %s", entry.unique_id)
+
+    if CONF_USERNAME not in entry.data:
+        LOGGER.debug("Super Old config entry format detected: %s", entry.unique_id)
+        raise ConfigEntryAuthFailed
+
+    if CONF_IP_ADDRESS not in entry.data:
+        LOGGER.debug("Old config entry format detected: %s", entry.unique_id)
+        entry = await _async_pseudo_migrate_entry(hass, entry)
+
+    # Fireplace will throw an error if it can't connect
+    try:
+        fireplace: UnifiedFireplace = (
+            await UnifiedFireplace.build_fireplace_from_common(
+                _construct_common_data(entry)
+            )
+        )
+        LOGGER.info("Waiting for Fireplace to Initialized")
+        await asyncio.wait_for(_async_wait_for_initialization(fireplace), timeout=600)
+    except TimeoutError as err:
+        raise ConfigEntryNotReady(
+            "Initialization of fireplace timed out after 10 minutes"
+        ) from err
+
+    # Construct coordinator
+    data_update_coordinator = IntellifireDataUpdateCoordinator(
+        hass=hass, fireplace=fireplace
     )
 
-    await coordinator.async_config_entry_first_refresh()
-    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
+    LOGGER.info("Await first refresh")
+    await data_update_coordinator.async_config_entry_first_refresh()
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = data_update_coordinator
+
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
     return True
 
 
+async def _async_wait_for_initialization(fireplace, timeout=600):
+    """Wait for a fireplace to be initialized."""
+    while (
+        fireplace.data.ipv4_address == "127.0.0.1" and fireplace.data.serial == "unset"
```

Comment:
Does the supporting library claim the IP-address is 127.0.0.1 before it's initialized? Why is this an and-condition, is it OK if `fireplace.data.ipv4_address == "127.0.0.1"` but `fireplace.data.serial != "unset"`?